File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    flag=False
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        self.dfs(root,subRoot)
        return self.flag
    def dfs(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]):
        if not root:
            return
        if root.val==subRoot.val:
            if self.check(root,subRoot):
                self.flag=True
                return
        self.dfs(root.left,subRoot)
        self.dfs(root.right,subRoot)
    def check(self, p:Optional[TreeNode], q:Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        if p.val!=q.val:
            return False
        return self.check(p.left,q.left) and self.check(p.right,q.right)

File: Data Structures & Algorithms/count-number-of-islands/submission-6.java
class Solution {
    int m,n;
    public int numIslands(char[][] grid) {
        this.m=grid.length;
        this.n=grid[0].length;
        int res=0;
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]=='1')
                {
                    dfs(i,j,grid);
                    res++;
                }
            }
        }
        return res;
    }
    public void dfs(int r, int c, char[][] grid)
    {
        if(r<0 || c<0 || r>=m || c>=n || grid[r][c]=='0') return;
        grid[r][c]='0';
        dfs(r-1,c,grid);
        dfs(r+1,c,grid);
        dfs(r,c+1,grid);
        dfs(r,c-1,grid);
    }
}
